keyboards/cipulot/common/ec_switch_matrix.c

```diff
@@ -0,0 +1,304 @@
+                    }
+                }
+            }
+        }
+        wait_ms(5);
+    }
+
+    // Average the noise floor
+    for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+        for (uint8_t col = 0; col < MATRIX_COLS; col++) {
+            ec_config.noise_floor[row][col] /= DEFAULT_NOISE_FLOOR_SAMPLING_COUNT;
+        }
+    }
+}
+
+// Scan key values and update matrix state
+bool ec_matrix_scan(matrix_row_t current_matrix[]) {
+    bool updated = false;
+
+    for (uint8_t amux = 0; amux < AMUX_COUNT; amux++) {
+        disable_unused_amux(amux);
+        for (uint8_t col = 0; col < amux_n_col_sizes[amux]; col++) {
+            for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+                uint8_t adjusted_col        = amux == 0 ? col : col + amux_n_col_sizes[amux - 1];
+                sw_value[row][adjusted_col] = ec_readkey_raw(amux, row, col);
+
+                if (ec_config.bottoming_calibration) {
+                    if (ec_config.bottoming_calibration_starter[row][adjusted_col]) {
+                        ec_config.bottoming_reading[row][adjusted_col]             = sw_value[row][adjusted_col];
+                        ec_config.bottoming_calibration_starter[row][adjusted_col] = false;
+                    } else if (sw_value[row][adjusted_col] > ec_config.bottoming_reading[row][adjusted_col]) {
+                        ec_config.bottoming_reading[row][adjusted_col] = sw_value[row][adjusted_col];
+                    }
+                } else {
+                    updated |= ec_update_key(&current_matrix[row], row, adjusted_col, sw_value[row][adjusted_col]);
+                }
+            }
+        }
+    }
+
+    return ec_config.bottoming_calibration ? false : updated;
+}
+
+// Read the capacitive sensor value
+uint16_t ec_readkey_raw(uint8_t channel, uint8_t row, uint8_t col) {
+    uint16_t sw_value = 0;
+
+    // Select the multiplexer
+    select_amux_channel(channel, col);
+
+    // Set the row pin to low state to avoid ghosting
+    writePinLow(row_pins[row]);
+
+    ATOMIC_BLOCK_FORCEON {
+        // Set the row pin to high state and have capacitor charge
+        charge_capacitor(row);
+        // Read the ADC value
+        sw_value = adc_read(adcMux);
+    }
+    // Discharge peak hold capacitor
+    discharge_capacitor();
+    // Waiting for the ghost capacitor to discharge fully
+    wait_us(DISCHARGE_TIME);
+
+    return sw_value;
+}
+
+// Update press/release state of key
+bool ec_update_key(matrix_row_t* current_row, uint8_t row, uint8_t col, uint16_t sw_value) {
+    bool current_state = (*current_row >> col) & 1;
+
+    // Normal board-wide APC
+    if (ec_config.actuation_mode == 0) {
+        if (current_state && sw_value < ec_config.rescaled_mode_0_release_threshold[row][col]) {
+            *current_row &= ~(1 << col);
+            uprintf("Key released: %d, %d, %d\n", row, col, sw_value);
```

Comment:
Perhaps done somewhere else and i didn't see it... But i would guard all the prints such that end-users have to explicitly enable the debugging options (eg see how `qp_dprintf` is defined)

Reply:
I want the prints to be always available since they give critical information on the press/release of the keys and give visual feedback with HID consoles for debug but also general use for showcase.

Reply:
really should be dprintf.

Reply:
the print is supposed to happen always, regardless of the user enabling or not debug mode, hence the use of `uprintf` instead of `dprintf`.